### PR TITLE
(upgrade) Benchmark sim

### PR DESCRIPTION
**Pull request by sneakyweasel**
- Added full level performance benchmarks
- Added "yarn benchmark" shortcut
- Added two benchmark levels from QG2
- Rewrote tsconfig JSON to work with ts-node
- Rewrote an example benchmark in TS working with benchmark.js suite

### CODE REVIEW

**Comment by sneakyweasel** [comment]
Build is not passing because of tsconfig:
Error: Incompatible tsconfig option. Module resolves to 'CommonJS'. This is incompatible with rollup, please use 'module: "ES2015"' or 'module: "ESNext"'".
Either we have CommonJS and we can run files with ts-node or we build with rollup. I'll try to see if there's another solution.

**Comment by stared** [comment]
@sneakyweasel Did you test the package?

Changes in Rollup may create a package that does not work with other tools (e.g. only browser, but not as an NPM package, or vice versa).

File: benchmarks/benchmark.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+sim.initializeFromIndicator(laserIndicator)
+sim.generateFrames(100)
+console.timeEnd('100 frames')
+
+console.time('Hall of mirrors')
+sim = new Simulation(grid2)
+laserIndicator = sim.generateLaserIndicator()
+sim.initializeFromIndicator(laserIndicator)
+sim.generateFrames(100)
+console.timeEnd('Hall of mirrors')
```
Comment (Frizi) [inline benchmarks/benchmark.ts:17]: Running stuff once in JS is a not a good measure of performance, as the compiler is being busy recompiling the hot paths in background, and many functions are not yet flagged for optimization, but would be after a few more runs.

To do this properly, the easiest way is to use an benchmarking framework. I see other bench suite (testMul) is using "benchmark" package, why have you decided to not use it here?

Also I don't have any favourites here, but if you are looking for a change, then https://github.com/caderek/benny seems like a nice typescript layer on top of "benchmark" with great defaults.

**Review by Frizi — CHANGES_REQUESTED** [review]

File: tsconfig.json [hunk 2]
```diff
@@ -3,21 +3,22 @@
...
     /* Basic Options */
     // "incremental": true,                   /* Enable incremental compilation */
-    "target": "es5",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
-    "module": "es2015",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "target": "ESNext",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "ESNext",                       /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    // "module": "CommonJS",                  /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
     "lib": [                                  /* Specify library files to be included in the compilation. */
-      "es2019",
+      "ESNext",
```
[the remark this answers is not in the document]
Comment (sneakyweasel) [reply tsconfig.json:10]: We can use a locked version if you prefer.

[the remark this answers is not in the document]
Comment (sneakyweasel) [reply tsconfig.json:10]: The main "problem" here is needing modules to be CommonJS to run ts-node and that RollupJS doesn't like it. So to run tests you must make this manual change. I didn't figure a workaround.

File: benchmarks/benchmark.ts (hunk 1 above)
Comment (sneakyweasel) [reply benchmarks/benchmark.ts:17]: It was to show two different benchmarking options: a "naive" one and the other using the BenchmarkJS framework.
BenchmarkJS seems to be industry standard with 4k stars compared to benny but wasn't updated for some time. Both have very similar syntax.

Comment (sneakyweasel) [reply benchmarks/benchmark.ts:17]: It feels the benchmarkJS / benny packages are more tailored to test function implementations and not full simulation runs, the output Event is:
- Run 1st benchmark level x 0.77 ops/sec ±5.28% (6 runs sampled)

Comment (Frizi) [reply benchmarks/benchmark.ts:17]: Benny is just a wrapper over BenchmarkJS. Use whichever you like more.

The output is indeed a bit cryptic once you get into such a long running functions like our simulation, but the information is there, and I believe the format is configurable.
If you are able to change the formatting, I would prefer it like:

- Hall of mirrors: 1298.70ms ±68.57ms (6 runs sampled)

That way we see the raw linear speed, not the inverse of it

File: tsconfig.json [hunk 3]
```diff
@@ -3,13 +3,13 @@
     "skipLibCheck": false,
     /* Basic Options */
     // "incremental": true,                   /* Enable incremental compilation */
-    "target": "ESNext",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
-    "module": "ESNext",                       /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "target": "ES6",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "ES6",                       /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
```
Comment (Frizi) [inline tsconfig.json:7]: "module" option describes used module syntax (are those es2015 modules with imports, commonjs requires or what else), while "target" describes the assumed level of features at the target runtime (say an old browser would require ES3)

With that in mind, i believe we should have `"target: "ES2019", "module": "es2015"`.

I believe the issue you had is related to `"ES6"` not being a valid module format, so it defaults to `"commonjs"`.

Comment (sneakyweasel) [reply tsconfig.json:7]: At least on my computer, I can't run ts-node used to run the benchmarks without having "module": "commonJS" enabled. I get the infamous: SyntaxError: Cannot use import statement outside a module. Any clue?

Comment (sneakyweasel) [reply tsconfig.json:7]: Target has those available values from VSCode:

- Specify ECMAScript target version: 'ES3', 'ES5', 'ES6'/'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', 'ESNext'

- Specify module code generation: 'None', 'CommonJS', 'AMD', 'System', 'UMD', 'ES6', 'ES2015', 'ES2020' or 'ESNext'. Only 'AMD' and 'System' can be used in conjunction with --outFile.

Comment (Frizi) [reply tsconfig.json:7]: try adding `"type": "module"` to `package.json`

**Comment by sneakyweasel** [comment]
- Switched from benchmarkJS to benny for simple function benchmark / performance comparisons.
- Implemented the performance NodeJS tools https://nodejs.org/api/perf_hooks.html which are better for full simulation runs imo.

**Review by Frizi — APPROVED** [review]
I still don't like that `benchmarks/benchmark.ts` is being called a benchmark, while it's just a few freestanding function invocations without any repetition. But I believe we can fix that as a separate PR, as there are many other valuable things waiting. Let's merge that and make that benchmark real separately.

**Pull request merged**